### PR TITLE
Relaxing `setv+` / `let+` / etc

### LINKED ISSUE #27: `let+` / `setv+` raising instead of giving `None`
Here are some calls to `let+`, to illustrate its current behavior:

```hy
(require hyrule.destructure [let+])

;;; List Pattern: simple
(let+ [[foo]
       ["quux"]]
  (print foo)) ; => "quux"

;;; List Pattern: ignoring type
(let+ [[foo]
       (, "quux" "frob")]
  (print foo)) ; => "quux"

;;; List Pattern: ignoring size / existence
(let+ [[foo]
       []]
  (print foo)) ; => None



;;; Dict Pattern: simple
(let+ [{foo :bar}
       {:bar "quux"}]
  (print foo)) ; => "quux"

;;; Dict Pattern: inexistent key
(let+ [{foo :frob}
       {:bar "quux"}]
  (print foo)) ; => None

;;; Dict Pattern: ignoring size / existence
(let+ [{foo :bar}
       {}]
  (print foo)) ; => None



;;; Dict Pattern: nested
(let+ [{{foo :bar} :baz}
       {:baz {:bar "quux"}}]
  (print foo)) ; => "quux"

;;; Dict Pattern: nested and ignoring size / existence
(let+ [{{foo :bar} :baz}
       {:baz {}}]
  (print foo)) ; => None
```

From there, a quick pattern emerges: the destructuring macros try to match, and if they can't, the variables are bound to `None`. That happens even if there is a type mismatch, a missing key, or if the size / structure doesn't fit perfectly. So destructuring is a `None`-ing / [Option<T>/Maybe<T>](https://en.wikipedia.org/wiki/Option_type) operation instead of a raising one (on the lines of ["ask for forgiveness" vs "ask for permission"](https://stackoverflow.com/a/12265860/6141224)).

However, given enough complex input, the `None`/optional quality of `let+` breaks:

```hy
(require hyrule.destructure [let+])

;;; Dict Pattern: nested
(let+ [{{foo :bar} :baz}
       {:baz {:bar "quux"}}]
  (print foo)) ; => "quux"

;;; Dict Pattern: nested and ignoring size / existence
(let+ [{{foo :bar} :baz}
       {:baz {}}]
  (print foo)) ; => None

;;; Dict Pattern: nested and ignoring size / existence
(let+ [{{foo :bar} :baz}
       {}]
  (print foo)) ; AttributeError: 'NoneType' object has no attribute 'get'



;;; Dict Pattern: nested and ignoring type
(let+ [{{foo :bar} :baz}
       {:baz []}]
  (print foo)) ; AttributeError: 'list' object has no attribute 'get'

;;; Dict Pattern: nested and ignoring size / existence (:or doesn't fix it)
(let+ [{{foo :bar} :baz :or {foo None}}
       {}]
  (print foo)) ; AttributeError: 'NoneType' object has no attribute 'get'

;;; Dict Pattern: nested and ignoring size / existence (:or doesn't fix it)
(let+ [{{foo :bar :or {foo None}} :baz}
       {}]
  (print foo)) ; AttributeError: 'NoneType' object has no attribute 'get'
```

This is really unfortunate because code that needs that the destructuring macros match some data very strictly (or otherwise raise an error) will be riddled with checks for `None` and type dismissiveness. On the other hand, code that only needs to vaguely match some structure will be riddled with `try`/`except` (or fallback to manually write the `if`s and `get`s that the destructuring macros are supposed to replace).

This really seems like a bug, we either should be more permissive with the matches (which seems like the fix) or not permissive at all. Here is some code that should be perfectly fine but is broken due to the current implementation (the breaking lines are commented out, but they are needed in order for the correct behavior of the program):

```hy
(require hyrule.destructure [let+])

;; Here is some complex json objects that could have come from
;; the database of a big store/shop website:
(setv
  cap {:name "eight-point cap"
       :type "headgear"
       :covers "head"}
  tank-top {:name "tank-top"
            :type "shirt"
            :coverage-data {:covers ["torso"] :percentage 0.80}}
  pants-socks-n-shoes {:name "pants, socks and shoes"
                       :type "kit (sale)"
                       :kit-covers ["legs" "feet"]}
  magazine {:name "sew awesome monthly" :type "magazine"}
  store-db [cap tank-top pants-socks-n-shoes magazine])

(defn covers-something? [item]
  (let+ [{covers1 :covers} item                   ; Single coverage item
         {covers2 :kit-covers} item               ; Multiple coverage item
         ;{{covers3 :covers} :coverage-data} item  ; Partial coverage item
         ]
    (cond
      [covers1 [covers1]]
      [covers2 covers2]
      ;[covers3 covers3]
      [:else False])))

;; Can you wear a cap?
(if (covers-something? cap)
    (print "You can wear it.")
    (print "It doesn't seem wearable.")) ; => You can wear it.

;; What body parts would it cover?
(print (covers-something? cap)) ; => ["head"]

;; Can you wear a pants-socks-n-shoes?
(if (covers-something? pants-socks-n-shoes)
    (print "You can wear it.")
    (print "It doesn't seem wearable.")) ; => You can wear it.

;; What body parts would it cover?
(print (covers-something? pants-socks-n-shoes)) ; => ["legs", "feet"]

;; Can you wear a magazine?
(if (covers-something? magazine)
    (print "You can wear it.")
    (print "It doesn't seem wearable.")) ; => It doesn't seem wearable.

;; The following two are broken due to the deconstruction macros:

;; What body parts would a tank-top cover?
(covers-something? tank-top) ; => False (!!)
                             ; should be => ["torso"]

;; What are the wearable items available?
(lfor item store-db
      :if (covers-something? item)
      (get item :name)) ; => doesn't have "tank-top" (!!)
                        ; should be => ["eight-point cap" "tank-top" "pants, socks and shoes"]
```

### PR DESCRIPTION
**Pull request by gabriel-francischini**
This fixes #27 by setting failing variables to `None` instead of raising an error on the first failed-to-match variable.
This pull request affects pretty much all destructuring macros.

Previous behavior:
```hy
(require hyrule.destructure [let+])

(let+ [{{meats :meats} :foods}
       {:foods {"nothing" "here"}}]
  (print meats)) ; => None

;; It's always possible to find some spam
(let+ [{{meats :meats :or {meats ["spam"]}} :foods}
       {"nothing" "here"}]
  (print meats)) ; AttributeError: 'NoneType' object has no attribute 'get'
```

Behavior after this pull request:
```hy
(require hyrule.destructure [let+])

(let+ [{{meats :meats} :foods}
       {:foods {"nothing" "here"}}]
  (print meats)) ; => None

;; It's always possible to find some spam
(let+ [{{meats :meats :or {meats ["spam"]}} :foods}
       {"nothing" "here"}]
  (print meats)) ; => ["spam"]
```

### CODE REVIEW

**Comment by Kodiologist** [comment]
I haven't made any real use of the destructing stuff, but wouldn't you more often want an exception to be thrown, instead of returning `None`? That's how Python's own unpacking works, as in `a, b = foo`, which raises an error if `foo` is not iterable or doesn't have exactly 2 elements.

It's a tiny thing, but we traditionally write `(is foo None)` instead of `(= foo None)`.

**Comment by allison-casey** [comment]
`(is foo None)` is not just tradition, `(= foo None)` is a pep8 error :woman_shrugging: 

This might also be an opportunity for us. Now that python3.10 `match` has made a canonical python syntax for destructuring it might be a worth changing the destructuring syntax to be directly aligned with Python. 

I've also been meaning to make destructuring favor string keys like we've been moving to with `(:akeyword {})` looking up by string. The current implementation and documentation pretty heavily encourages using keywords and not strings

**Comment by gabriel-francischini** [comment]
You're right, for a moment I forgot that `is` was a thing and mistakenly thought that the only options were `(if foo ...)` and `(if (= foo None)`. 

About favoring strings, `(:foo {"foo" 10})` resulting in `10` was a nice surprise when I upgraded Hy from 0.20 to the latest version. That simplifies dealing with python dictionaries, especially given the use of strings instead of keywords (due to python not having keywords). So these 3 do look somewhat bug-ish:

```hy
(do (setv+ {x :foo} {"foo" 10})
    (print x)) ; => None
(get {"foo" 10} :foo) ; => KeyError: hy.models.Keyword('foo')
(.get {"foo" 10} ':foo) ; => None
```

Favoring strings on the first one seems like the right thing to do. The last one is probably out of our control, and I don't have any particular opinion. 

Maybe `{:foo 1 "foo" 2}` should also be converted to `{"foo" 1 "foo" 2}`? I don't know, but favoring (or at least accepting) strings in dictionary patterns seems like it would avoid some bugs/confusions due to string-vs-keyword differences, and would also simplify code that destructures dictionaries produced by python libraries.

**Comment by Kodiologist** [comment]
There was a point where Hy keywords actually evaluated to strings, rather than themselves. We've struggled over time with how exactly to draw the distinctions between them. My sense now is that things generally work out better if you treat them as different, and expect the programmer to use strings in many cases where they would use keywords in other Lisps. Although this is counterintuitive to people coming from other Lisps, it gets us the useful keywords-evaluate-to-themselves semantics while still playing nicely with Python libraries, which expect strings.

I can see why you might want a literal keyword in a literal dictionary to magically be interpreted as a string instead, but consider that you can already get syntax like this with `(dict :foo 1)`. As in Python, you usually make dictionaries with `dict`, and resort to dictionary literals only when you need finer control over the keys.

**Comment by gabriel-francischini** [comment]
That's interesting, I didn't know that `dict(foo=1)` produces a dict like that. I don't remember that being mentioned in the docs (especially in the tutorial), although it seems like common sense once you know it.

Since differentiating between them will be important moving forwards, I think that the destructuring code should also be able to differentiate between them, so I'll leave that part as it is. 

Also, maybe we should point in the docs (perhaps on the tutorial) ways to quickly convert between keyword-dicts and string-dicts (like `(dict :foo 1)`)? On that note, I don't know why `(dict :foo 1)` works but `(dict #** {:foo 1})` doesn't.

Maybe I'm overthinking this, a dictionary comprehension (`dfor`) should be able to convert between the two, although not recursively. There has to be a simple way to convert between the two without reinventing the wheel every time, I just don't know it yet.

**Comment by allie-jo** [comment]
> That's interesting, I didn't know that dict(foo=1) produces a dict like that. I don't remember that being mentioned in the docs (especially in the tutorial), although it seems like common sense once you know it.

The tutorial definitely needs a re-write and it's on my todo list once we get through the last of the bigger changes.

Using keywords as dictionary keys is actively discouraged. It's only allowed because any hashable data type can be used as a dictionary key. It's not in Hy's scope to provide a way to convert between what is the encouraged canonical python form and what is effectively an anti-pattern.

**Comment by allie-jo** [comment]
In my opinion the destructuring syntax should not have both `:keys` and `:strs`. There should only be `:keys` which does what `:strs` currently does. There should be no syntax dedicated to destructuring keywords from dictionaries

**Comment by Kodiologist** [comment]
> Using keywords as dictionary keys is actively discouraged. It's only allowed because any hashable data type can be used as a dictionary key.

I don't find it hard to imagine some legitimate uses for keyword objects as dictionary keys, as when, for example, you have a complicated macro whose behavior is controlled by keywords, and you're looking up what to do with each keyword in a dictionary. It's just that keywords as dictionary keys is not what you really want most of the time.

I can't speak for the issue of whether `:keys` and `:strs` are both worth having for destructuring, although admittedly the value of the keyword case seems limited.

**Comment by Kodiologist** [comment]
> The tutorial definitely needs a re-write

I'm reasonably proud of it, actually, and I wouldn't encourage throwing out the whole thing and starting over, but I'm sure there are many improvements one could make.

**Comment by allie-jo** [comment]
> I don't find it hard to imagine some legitimate uses for keyword objects as dictionary keys

Sorry, i was unclear. I think destructuring should be able to pull values from any valid hashable key (including keywords), but there shouldn't be a dedicated syntax for pulling values out via keywords as `:keys` currently does. 

>  I wouldn't encourage throwing out the whole thing and starting over

Oh i wouldn't throw out the whole thing by any means. I would more like to lean into some of the decisions we've made over the last year. Make more things explicit than anything else. That and reorganize the docs as a whole a bit.

**Comment by Kodiologist** [comment]
> reorganize the docs as a whole a bit

Definiitely needed. It's something I've been meaning to get to, to, so it's just a question of which of us gets the tuits first.

**Comment by allie-jo** [comment]
Just to go back a sec to potentially changing the destructuring syntax to align with the destructuring syntax used by 3.10 `match`. Here's what I had in mind.

```clojure
(let+ [{"a key" a-sym  {"a" [{:keys [b c d]} :as nested  e f]}} some-big-dict]
  [a-sym b c d nested e f])
```
This has key and binding symbol swapped compared to clojure, but Hy isn't clojure 🤷‍♀️  and I like the consistent syntax with our existing `match` implementation. the only addition here is the `:keys` option. Which, looks up by string keys and maintains the idiom of keywords being arguments not identifiers themselves (even though there's nothign stopping you from destructuring keywords here or in `match`

**Comment by gabriel-francischini** [comment]
Using keywords in dictionary keys may be an anti-pattern in Hy, but I think that leaves a lot on the table just to be able to interop with python. Keywords are data type designed to represent keys, while strings are designed to represent text and gets hijacked to represent keys.

If you're using only Hy code, it seems reasonable to use the best data structure to represent the thing at hand — keywords for keys and strings for text (and never the other way around). Having `(:keyword dict)` encourages that, so maybe we shouldn't have that syntax?

**Comment by Kodiologist** [comment]
We got pretty distracted by this keyword issue, but I'd still like you to consider this question, @gabriel-francischini:

> I haven't made any real use of the destructing stuff, but wouldn't you more often want an exception to be thrown, instead of returning `None`? That's how Python's own unpacking works, as in `a, b = foo`, which raises an error if `foo` is not iterable or doesn't have exactly 2 elements.

**Comment by gabriel-francischini** [comment]
About that question, usually when I need destructuring I'm either matching (like the recently added python `match`) or filtering for certain data structures.

A comparison with python's `re` library is very reasonable. Regex is about manipulating text, and destructuring is about manipulating data structures. Most `re` functions are non-raising and returns some variation of `None` when failing, like [`re.match`](https://docs.python.org/3/library/re.html#re.match) (returns `None` on fail, otherwise a `MatchObject`), `re.sub` (returns original string on fail, otherwise a new string) and `re.findall` (returns `[]` on fail).

As you can see, if you should raise or not and what kind of value you should return in case of a fail depends on the specifics of the use case. With a implementation returing `None` in the general case, the user can write a few `if`s and do the right thing. With specific enough calls, like `defn+`, `fn+` and some hypothetical `branch+` or `and+` we can figure out the right thing to do, but in the general case, if we are going to have only one implementation, I think we should go with a `None`-ing one  (because it's easier to add some code `raise` from a `None`-ing implementation, but to recover the bindings from a `raise`-ing implementation is not so easy).

That's why you can get away with raising when unpacking tuples or iterables, it's specific enough that you can know that failure-to-match is a programming error.

Python used to allow [unpacking in function arguments](https://www.python.org/dev/peps/pep-3113/), but they removed it. Our `defn+` does what that used to do.

**Comment by Kodiologist** [comment]
Okay, good explanation.

**Comment by Kodiologist** [comment]
Anyway, I read it over. This looks pretty good. Rebase to squash the fix-up commit into the commit you're fixing; add the URL of the issue as a comment at the top of the test; and don't forget about AUTHORS, NEWS, and the documentation (see Hy's `CONTRIBUTING.rst`).

**Comment by gabriel-francischini** [comment]
Right. I'm considering that `None`-ing in the case of `defn+` and `fn+` is akin to silently ignoring errors on `a, b = foo`. I suspect some of this error suppression was already present before these changes. This PR should be backward-compatible though, so we are not making things worse than they were before (for existing code, at least).

Once this PR goes through, I can add a PR with a `raise`-ing implementation that fixes those error suppressions and provides proper flexibility for users/use-cases that want to `raise` in the case of a failure-to-match (which should be a minority, but an important one as `defn+` and `fn+` shows).

I'll do the changes you mentioned, @Kodiologist .

**Comment by Kodiologist** [comment]
Thanks. For the documentation, I would say just move the sentence "Variables which are not found in the expression are set to ``None`` if no default value is specified" up and maybe rephrase it a bit, since it should apply to all sorts of patterns.

**Comment by gabriel-francischini** [comment]
Sure. I'll try do add that phrase and an example, to drive the point home since that can silently suppress warnings if not paid attention to. I'll also the issue URL in the tests and fix a newline issue on `AUTHROS.rst`

**Comment by gabriel-francischini** [comment]
@Kodiologist Done.

Users can currently use the `:as` pattern to selectively check which variables have successfully matched and which haven't. 

That still is some sort of a hack, but it's a reasonable compromise until we have both this non-raising implementation and a raising one (and a way to expose the bindings, so that users can write macros that satisfy their specific destructuring needs without needing to dive into `hyrule.destructure`'s code).

**Review by Kodiologist — APPROVED** [review]

**Comment by Kodiologist** [comment]
@allison-casey Did you want to review this? Otherwise, it's clear to merge.

**Review by allison-casey — APPROVED** [review]
